modules/tools/rosbag/body_sensation_evaluation.py

```diff
@@ -0,0 +1,224 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###############################################################################
+"""
+Usage:
+    python body_sensation_evalution.py bag1 bag2 ...
+"""
+
+import collections
+import math
+import sys
+import time
+
+from rosbag.bag import Bag
+
+from modules.canbus.proto.chassis_pb2 import Chassis
+
+kChassisTopic = '/apollo/canbus/chassis'
+kLocalizationTopic = '/apollo/localization/pose'
+
+class BodySensationCalculator(object):
+
+    def __init__(self):
+        self.driving_mode = []
+        self._timestamp = 0.0
+        self._bump_lock_time = 0.0
+        self._last_bump_time = 0.0
+        self._last_speed_down_2_time = 0.0
+        self._last_speed_down_4_time = 0.0
+        self._last_speed_up_2_time = 0.0
+        self._last_speed_up_4_time = 0.0
+        self._last_turning_time = 0.0
+        self._speed_down_2_flag = 0
+        self._speed_down_4_flag = 0
+        self._speed_up_2_flag = 0
+        self._speed_up_4_flag = 0
+        self._turning_flag = 0
+        self.auto_counts = {}
+        self.auto_counts["speed_down_2"] = 0
+        self.auto_counts["speed_down_4"] = 0
+        self.auto_counts["speed_up_2"] = 0
+        self.auto_counts["speed_up_4"] = 0
+        self.auto_counts["turning"] = 0
+        self.auto_counts["bumps"] = 0
+        self.manual_counts = {}
+        self.manual_counts["speed_down_2"] = 0
+        self.manual_counts["speed_down_4"] = 0
+        self.manual_counts["speed_up_2"] = 0
+        self.manual_counts["speed_up_4"] = 0
+        self.manual_counts["turning"] = 0
+        self.manual_counts["bumps"] = 0
+
+    def get_driving_mode(self, bag_file):
+        """get driving mode, which is stored in a dict"""
+        with Bag(bag_file, 'r') as bag:
+            mode={}
+            mode["status"] = 'UNKNOW'
+            mode["start_time"] = 0.0
+            mode["end_time"] = 0.0;  
+            for topic, msg, _t in bag.read_messages([kChassisTopic]):
+                t = long(str(_t))*pow(10,-9)
+                cur_status = msg.driving_mode
+                if mode["status"] != cur_status:
+                    if mode["status"] != 'UNKNOW':
+                        self.driving_mode.append(mode)
+                    mode["status"] = cur_status
+                    mode["start_time"] = t
+                mode["end_time"] = t
+            self.driving_mode.append(mode)
+
+    def _check_status(self, timestamp):
+        """check driving mode according to timestamp"""
+        for mode in self.driving_mode:
+            if timestamp >= mode["start_time"] and timestamp <= mode["end_time"]:
+                if mode["status"] == Chassis.COMPLETE_AUTO_DRIVE:
+                    return True
+                else:
+                    return False
+        return False
+
+    def _bumps_rollback(self, bump_time):
```

Comment:
use bumps_rollback_ instead

Reply:
Could you explain why? This method should be protected.

Reply:
sorry my bad here